### PR TITLE
Allow group guide rates in case that a group is an auto choke group

### PR DESCRIPTION
**Pull request by plgbrts**
This PR is to allow group guide rates in case that a group is an auto choke group:
In the example case we have:
```
GCONPROD--  1    2     3     4       5     6     7       8       9       10 
'M5S' ORAT  8000  1*      1*    1*  'RATE'  'YES'   40.0   'OIL'    1*   1*   1*   3* /
'B1'  FLD  1*     1*      1*    1*  'RATE'  'YES'   35.0   'OIL'    1*   1*   1*   3* /
'C1'  FLD  1*    1*       1*    1*  'RATE'  'YES’   5.0   'OIL'    1*   1*   1*   3* / 
```
and
```
NODEPROP
--  NodeName  Press  Choke
    PLAT-A    21.0   NO   /
    M5S       1*     NO   /     
    B1        1*     YES  / 
    C1        1*     NO   /
```
The `M5S` group is an ancestor of the production groups `B1` and `C1` and `B1` is an autochoke group ( https://github.com/OPM/opm-simulators/pull/4935).  The target rates of B1 and C1 become `7000 (=8000*35/40)` and `1000 (=8000*5/40)`
A comparison with a reference simulator is shown in the figure
image
Note that OPM-Flow keeps wells open for a longer time.
The test case is: [GROUPGUIDERATES_AC_8000_OPM_TEST.zip](https://github.com/user-attachments/files/17846956/GROUPGUIDERATES_AC_8000_OPM_TEST.zip)

### CODE REVIEW

**Comment by plgbrts** [comment]
jenkins build this opm-common=4355 please

**Comment by plgbrts** [comment]
jenkins build this opm-common=4355 please

**Comment by plgbrts** [comment]
jenkins build this opm-common=4355 please

File: opm/simulators/wells/WellGroupControls.cpp [hunk 1]
```diff
@@ -549,6 +556,116 @@ getGroupProductionTargetRate(const Group& group,
     return scale;
 }
 
+template<class Scalar>
+std::pair<Scalar, Group::ProductionCMode> WellGroupControls<Scalar>::
+getAutoChokeGroupProductionTargetRate(const std::string& name,
+                                      const Group& group,
```
Comment (totto82) [inline opm/simulators/wells/WellGroupControls.cpp:562]: What is the difference between this code and getGroupProductionTargetRate? Can this be refactored to avoid code duplication.

**Review by totto82 — CHANGES_REQUESTED** [review]
Thanks for the contribution. I have done the first review. I still struggle to understand the complete logic here, but for what I understand the code looks ok. My main concern is the code duplication in getAutoChokeGroupProductionTargetRate. This code is complicated so for maintenance it would be good to reduce the places the logic is copied. Also, I am not fully convinced that using guiderates from the production control is always ok. What if it is defaulted? Shouldn't it then accumulate it from its children group/wells?

**Comment by plgbrts** [comment]
jenkins build this opm-common=4355 please

File: opm/simulators/wells/BlackoilWellModel_impl.hpp [hunk 2]
```diff
@@ -1317,19 +1318,51 @@ namespace Opm {
...
                                                   gratTargetFromSales, nodeName, group_state,
                                                   group.has_gpmaint_control(cmode));
-                const Scalar orig_target = tcalc.groupTarget(ctrl, local_deferredLogger);
+                if (!fld_none)
+                {
+                    target_tmp = tcalc.groupTarget(ctrl, local_deferredLogger);
+                }
+
+                const Scalar orig_target = target_tmp;
+                std::cout<< "Group: " << group.name() << " orig_target: " << orig_target*86400 << std::endl;
```
[the remark this answers is not in the document]
Comment (plgbrts) [reply opm/simulators/wells/BlackoilWellModel_impl.hpp:1388]: The `target_tmp` cannot be `const` as it is set conditionally. I introduced the `const` `orig_target` variable to ensure its  the value cannot change.

File: opm/simulators/wells/FractionCalculator.cpp [hunk 3]
```diff
@@ -147,7 +150,7 @@ guideRate(const std::string& name,
         return WellGroupHelpers<Scalar>::getGuideRate(name, schedule_, well_state_, group_state_,
                                                       report_step_, guide_rate_, target_, pu_);
     } else {
-        if (groupControlledWells(name, always_included_child) > 0) {
+        if ((groupControlledWells(name, always_included_child) > 0)) {
             if (is_producer_ && guide_rate_->has(name)) {
```
[the remark this answers is not in the document]
Comment (plgbrts) [reply opm/simulators/wells/FractionCalculator.cpp]: Removed it.

**Comment by plgbrts** [comment]
Note that I have included https://github.com/OPM/opm-simulators/pull/5768 in this PR

> Thanks for the contribution. I have done the first review. I still struggle to understand the complete logic here, but for what I understand the code looks ok. My main concern is the code duplication in getAutoChokeGroupProductionTargetRate. This code is complicated so for maintenance it would be good to reduce the places the logic is copied. Also, I am not fully convinced that using guiderates from the production control is always ok. What if it is defaulted? Shouldn't it then accumulate it from its children group/wells?

These comments are addressed below

File: opm/simulators/wells/WellGroupHelpers.cpp [hunk 4]
```diff
@@ -1107,16 +1120,65 @@ groupControlledWells(const Schedule& schedule,
...
+                                              group_state,
+                                              group.has_gpmaint_control(control_group_cmode));
+            auto deferred_logger = Opm::DeferredLogger();
+            const auto& control_group_target = tcalc.groupTarget(ctrl, deferred_logger);
+            // Target rate for the auto choke group
+            const Scalar target_rate = control_group_target * group_guide_rate / control_group_guide_rate;
+            const Scalar current_rate = tcalc.calcModeRateFromRates(rates);
+
+            if (current_rate < 0.99 * target_rate)
+                included = false;
```
[the remark this answers is not in the document]
Comment (plgbrts) [reply opm/simulators/wells/WellGroupHelpers.cpp]: It appears not necessary so I removed it

File: opm/simulators/wells/WellGroupHelpers.cpp [hunk 5]
```diff
@@ -1107,16 +1120,65 @@ groupControlledWells(const Schedule& schedule,
...
+
+            // Get the ancestor of the auto choke group that has group control (cmode != FLD, NONE)
+            const auto& control_group_name = control_group(group, group_state, report_step, schedule);
+            const auto& control_group = schedule.getGroup(control_group_name, report_step);
+            const auto& ctrl = control_group.productionControls(summary_state);
+            const auto& control_group_guide_rate = ctrl.guide_rate;
+            const auto& control_group_cmode = ctrl.cmode;
+
+            const auto& group_guide_rate = group.productionControls(summary_state).guide_rate;
+
```
[the remark this answers is not in the document]
Comment (plgbrts) [reply opm/simulators/wells/WellGroupHelpers.cpp:1154]: If for (one of) the children groups the the guide rate is defaulted the behavior is now the same as if no guide rates are provided for all three groups:

```
GCONPROD--  1    2     3     4       5     6     7       8       9       10 
'M5S' ORAT  8000  1*      1*    1*  'RATE'  'YES'   1*   ''    1*   1*   1*   3* /
'B1'  FLD  1*     1*      1*    1*  'RATE'  'YES'   1*   ''    1*   1*   1*   3* /
'C1'  FLD  1*    1*       1*    1*  'RATE'  'YES’   1*   ''    1*   1*   1*   3* / 
```
This behaviour is similar as the reference simulator. A comparison for the GCONPROD above:
image

File: opm/simulators/wells/WellGroupHelpers.cpp [hunk 6]
```diff
@@ -1107,16 +1120,65 @@ groupControlledWells(const Schedule& schedule,
...
+                                                                                  phase_pos,
+                                                                                  false);
+            }
+
+            // Get the ancestor of the auto choke group that has group control (cmode != FLD, NONE)
+            const auto& control_group_name = control_group(group, group_state, report_step, schedule);
+            const auto& control_group = schedule.getGroup(control_group_name, report_step);
+            const auto& ctrl = control_group.productionControls(summary_state);
+            const auto& control_group_guide_rate = ctrl.guide_rate;
+            const auto& control_group_cmode = ctrl.cmode;
```
[the remark this answers is not in the document]
Comment (plgbrts) [reply opm/simulators/wells/WellGroupHelpers.cpp:1151]: This is fixed. If the guide rate is defaulted it will be calculated from the children guide rates. Thus for 
```
GCONPROD--  1    2     3     4       5     6     7       8       9       10 
'M5S' ORAT  8000  1*      1*    1*  'RATE'  'YES'   1*    ''    1*   1*   1*   3* /
'B1'  FLD  1*     1*      1*    1*  'RATE'  'YES'   35.0   'OIL'    1*   1*   1*   3* /
'C1'  FLD  1*    1*       1*    1*  'RATE'  'YES’   5.0   'OIL'    1*   1*   1*   3* / 
```
The guide rate of `M5S` will be internally calculated as 40.

File: opm/simulators/wells/WellBhpThpCalculator.cpp [hunk 7]
```diff
@@ -1051,13 +1051,14 @@ bruteForceBracketCommonTHP(const std::function<Scalar(const Scalar)>& eq,
...
-    constexpr int sample_number = 1000; 
+    constexpr int sample_number = 1000;
     constexpr Scalar interval = 1E5; 
     Scalar eq_low = eq(min_thp);
     Scalar eq_high = 0.0;
     for (int i = 0; i < sample_number + 1; ++i) {
         max_thp = min_thp + interval * i;
         eq_high = eq(max_thp);
+        // std::cout << "max_thp: " << max_thp/1E5 << " eq_high: " << eq_high <<  " eq_low: " << eq_low << std::endl;
         if (eq_high * eq_low <= 0.) {
```
Comment (totto82) [inline opm/simulators/wells/WellBhpThpCalculator.cpp:1062]: remove

File: opm/simulators/wells/WellGroupHelpers.cpp [hunk 8]
```diff
@@ -1296,6 +1394,7 @@ checkGroupConstraintsProd(const std::string& name,
             // we add a factor here to avoid switching due to numerical instability
             const Scalar factor = 1.01;
             if (currentRateFraction > (guiderateFraction * factor)) {
+                std::cout << guided_group << " localCurrentRate(guided_group): " << localCurrentRate(guided_group) << chain[ii-1] << " localCurrentRate(chain[ii-1]) " << localCurrentRate(chain[ii-1]) << std::endl;
                 return std::make_pair(true, guiderateFraction / currentRateFraction);
```
Comment (totto82) [inline opm/simulators/wells/WellGroupHelpers.cpp:1398]: remove

File: opm/simulators/wells/WellInterface_impl.hpp [hunk 9]
```diff
@@ -1482,6 +1482,7 @@ namespace Opm
                 } else {
                     ws.trivial_target = true;
                 }
+                    
                 break;
```
Comment (totto82) [inline opm/simulators/wells/WellInterface_impl.hpp:1486]: Revert this change. It is good to reduce the number of files touched in the PR

**Review by totto82 — APPROVED** [review]
Thanks for the update. I only have some very minor clean-up suggestions that should be done before this can be merged.

**Comment by plgbrts** [comment]
> Thanks for the update. I only have some very minor clean-up suggestions that should be done before this can be merged.

The clean up suggestions have been done. Thanks a lot @totto82 for the very helpful discussions!

**Comment by plgbrts** [comment]
jenkins build this opm-common=4355 please
